spacy/matcher/matcher.pyx

```diff
@@ -225,11 +227,37 @@ cdef class Matcher:
             raise ValueError(Errors.E156.format())
         matches = find_matches(&self.patterns[0], self.patterns.size(), doclike, length,
                                 extensions=self._extensions, predicates=self._extra_predicates)
-        for i, (key, start, end) in enumerate(matches):
+        final_matches = []
+        pairs_by_id = {}
+        # For each key, either add all matches, or only the filtered, non-overlapping ones
+        for (key, start, end) in matches:
+            span_filter = self._filter.get(key)
+            if span_filter is not None:
+                pairs = pairs_by_id.get(key, [])
+                pairs.append((start,end,end-start))
+                pairs_by_id[key] = pairs
+            else:
+                final_matches.append((key, start, end))
+        for key, pairs in pairs_by_id.items():
+            span_filter = self._filter.get(key)
+            matched = [False for i in range(length)]
```

Comment:
It's hard to know whether to worry about efficiency here, because different problem sets have different characteristics. I think it's reasonable to make some small adjustments to get ahead of potential problems.

We have some very nested looping here that is likely to be a problem. I suggest we just make the `matched` check a C array that we allocate once outside of the `pairs_by_id` loop, something like:

```
cdef Pool tmp_pool = Pool()
matched = <char*>tmp_pool.alloc(length, sizeof(char))
```

We would use `memset` to reset the array to 0 in the `pairs_by_id.items` loop, and then replace the innermost loop where you're just setting to `True` with `memset` as well. We could also define an empty array and use it as a target for `memcmp` to replace the `all` loop. This would give us something like this:

```
cdef Pool tmp_pool = Pool()
matched = <char*>tmp_pool.alloc(length, sizeof(char))
empty = <char*>tmp_pool.alloc(length, sizeof(char))

for key, pairs in pairs_by_id.items():
    memset(matched, 0, length * sizeof(matched[0]))
    for (start, end, span_len) in sorted_pairs:
        assert 0 >= start < end # Defend against segfaults
        assert (end-start) == span_len
        # If no tokens in the span have matched
        if memcp(&matched[start], &empty[start], span_len * sizeof(matched[0])) == 0:
            final_matches.append((key, start, end)) 
            # Mark tokens that have matched.
            memset(&matched[start], 1, span_len * sizeof(matched[0]))
```

It's always a bit uncomfortable making this type of optimization before running the code and finding that it's actually a problem. But unfortunately we just have to stay a bit vigilant, especially in the matcher where the dynamics of different problems can be really different. It's easy to have some code that's not a problem on anything we test, and then on some problems it runs for 100 years. I think the code here is O(n^2) but perhaps even cubic (I don't know what the guarantees are about number of matches vs match length), which gets glacially slow if the bottom of the loop is setting into a Python list.

For future reference, if the sorting also becomes a problem, we can replace those with a C++ `vector[pair[int, int]]` I think. I'd rather not though as that gets really ugly, hopefully the sorting that's there is okay, although with the lambda it might still get slow on large lists, not sure. An intermediate construct that's still Python but would be a bit faster is to construct the list so that it sorts in the desired order by default, i.e. you apply the negatives etc and change the ordering, and then you sort that transformed list.

Reply:
Hm, why does it loop like that actually? Can we do:

```python
for (start, end, span_len, key) in sorted_pairs:
```

Reply:
And maybe it's possible to generalize this a bit outside of this particular component? I wrote a reimplementation of `util.filter_spans` for `SpanC` for the tokenizer here, but maybe this all can be part of a general `SpanC` util function somewhere? https://github.com/explosion/spaCy/blob/b81a89f0a94ce5a191720ba0eccff43667da6ba9/spacy/tokenizer.pyx#L290-L304

Reply:
Good point, although generalizing this type of code is often really hard. If we can't immediately see how to make that work, I think it's okay to just sprinkle in a bit of C to address the problem.